### ISSUE TITLE
Evaluate 8 MHz operation — ~2x the PHY ceiling, and it is what MM6108 is rated for

### ISSUE THREAD

**Issue by winson3QQ** [winson3QQ/Batman#39]
Came out of closing #36. With MCS8/9 ruled out as silicon that does not exist, **bandwidth is
the only remaining large lever.**

## The opportunity

The US regulatory table this chip loads already contains 8 MHz channels:

```
Deconstructing ch 12 (908000 kHz, 8 MHz) into primaries
Deconstructing ch 28 (916000 kHz, 8 MHz) into primaries
Deconstructing ch 44 (924000 kHz, 8 MHz) into primaries
```

| | 4 MHz (today) | 8 MHz |
|---|---|---|
| MCS7 SGI PHY | 16.65 Mbps | **~33.3 Mbps** |
| measured UDP | 11.0 / 9.4 Mbps | ? |

~32.5 Mbps is also the figure Morse quotes as the MM6108 maximum, so 8 MHz is simply how you
reach what the part is rated for. `MORSE_CAPS_8MHZ` exists as a firmware capability bit and
`rc.c` already builds rate tables up to 8 MHz.

Prior work went 2 MHz -> 4 MHz and got +28% down / +60% up (`docs/field-test-log.md`).
8 MHz has never been tried.

## What has to be checked first

- **Regulatory.** Confirm 8 MHz is permitted for the deployment region and that TX power is
  not derated. Note the firmware regdb has no TW entry and falls back to AU
  (`docs/pi5-rp1-bringup.md`).
- **Range.** Doubling bandwidth halves power spectral density — roughly 3 dB of link budget.
  For a MANET where range matters more than peak throughput this may be the wrong trade.
  Needs measuring at distance, not on the bench.
- **#34 gets worse, probably.** The Pi 500 already runs 2.5-7% MPDU FCS failures at 4 MHz.
  Wider channel, more noise captured, lower SNR per bit. **#34 should be understood before
  this is judged**, or the Pi 500 will make 8 MHz look worse than it is.
- **Primary 1 MHz channel index.** Going 2 -> 4 MHz already required
  `s1g_prim_1mhz_chan_index` to match on both ends or SAE fails (see `docs/upgrade-1.8.0.md`).
  8 MHz will need the same care, with a different derived index.

## Danger

Changing bandwidth means changing the channel on **both** ends. If the node does not come
back up there is currently **no wired lifeline** — `eth0` on the Pi 500 has no carrier, and
SSH to the node runs over the mesh being reconfigured.

Do not attempt this without either a wired path to manet01 or physical access to the card.
The safe order is the one used for the 4 MHz switch: change the node first, commit, reboot,
then align the Pi 500; revert over wired if it does not come back.

## Suggested sequence

1. Do #38 first (manet01 <-> manet02 baseline at 2/4 MHz). Two proper nodes give a clean
   reference and a safe pair to experiment on that does not involve the Pi 500.
2. Try 8 MHz between those two nodes, on the bench, with physical access.
3. Measure at distance before adopting — peak throughput is not the objective for a MANET.

Related: #33, #34, #36, #38.

**Comment by winson3QQ** [comment]
## Measured: 4 MHz vs 8 MHz mesh throughput (2026-09-17, manet01↔manet02)

Both nodes on 1.3.0-wsl.2, MM6108 HaLow, US, bench (nodes close, **signal −31…−36 dBm** — best-case link). Metric: `batctl tp` (batman-adv throughputmeter over the mesh — isolates the HaLow link, not the ethernet management path). 10 s per run.

| BW | Channel / freq | PHY (mapped VHT) | `batctl tp` (steady) |
|---|---|---|---|
| **4 MHz** (baseline) | ch40 / 922 MHz | 325 Mbit/s VHT-MCS7 (80 MHz map) | ~**6.7–7.0 Mbps** warm |
| **8 MHz** | ch44 / 924 MHz | 650 Mbit/s VHT-MCS7 (160 MHz map) | ~**11.0–13.3 Mbps**, avg **~11.6 Mbps** |

**Result: ~1.65× throughput** going 4→8 MHz. The PHY rate doubled exactly (325→650); real batman-adv throughput scaled ~1.65×, the shortfall from 2× being fixed per-frame/airtime/protocol overhead — expected.

### Method notes / honesty
- **`batctl tp` ramps**: at 4 MHz the first runs read low (0.5–2 Mbps) and settled to ~7 Mbps by runs 5–6; 8 MHz was steady from run 1 (10.9–13.3). Reported numbers are the steady/warm values.
- US 8 MHz channels (morse-regdb): ch12/908, ch28/916, **ch44/924** (chosen — nearest to the ch40 operating point).
- Switch = `uci set wireless.radio1.channel=44; commit; wifi reload` on both; regdb derives bw=8 from the channel. Mesh split then reformed in <25 s.

### Caveat — this is the throughput half only
This was measured at **−36 dBm (nodes adjacent)**. 8 MHz doubles the occupied bandwidth → **~3 dB worse receiver sensitivity and more captured noise → shorter range / less robustness** than 4 MHz for the same link budget. That cost is invisible at bench range. **The range/robustness half needs a field test** (coverage survey) before 8 MHz can be a default.

### Recommendation
Keep **4 MHz as the golden default** (robust, the field-validated config); treat **8 MHz as an opt-in for short-range / high-SNR links** where the ~1.65× throughput is worth the range trade. Revisit as a default only after a field range comparison (#39 range half + coverage survey). Related RF tuning still open: #37 (A-MPDU vs TBTT), #40 (beacon_int, currently 1000 ms).

**Comment by winson3QQ** [comment]
## Correction: use iperf, not `batctl tp` — `batctl tp` badly under-reads

The `batctl tp` numbers above under-report the link. Re-measured with **iperf (TCP, over the mesh path, RTT ~3.3 ms, signal −33 dBm)**:

| BW | `batctl tp` (under-read) | **iperf TCP (real)** |
|---|---|---|
| **4 MHz** (ch40/922) | ~7 Mbps | **~9.5 Mbps** (9.52 / 9.56) |
| **8 MHz** (ch44/924) | ~11.6 Mbps | **~15.9 Mbps** (+ UDP ~14 Mbps, 0% loss = sender-limited, not link-saturated) |

- **Ratio ≈ 1.67×** (unchanged conclusion; PHY doubles 325→650 Mbit/s VHT-MCS7, real gain ~1.67× from fixed overhead).
- The 4 MHz iperf figure (~9.5 Mbps) matches the previously-remembered ~10 Mbps result — confirming iperf is the right yardstick and `batctl tp` (batman's own tp-meter, with its ramp + protocol overhead) reads ~25–30% low here.

Method: `iperf -s` on one node, `iperf -c <mesh-ip> -t 10` on the other; path confirmed over br-ahwlan/bat0 (HaLow), not the ethernet management link. 8 MHz TCP was a single 10 s run; 4 MHz was two. Nodes reverted to 4 MHz/ch40 (golden default) after the test. Range/robustness half still untested (see caveat above).

**Comment by winson3QQ** [comment]
## 8 MHz 5-min soak charts (manet02→manet01, OTS running on manet01 throughout)

Same panel layout as the 4 MHz soak: throughput / CPU (both nodes) / RAM (both) / RSSI rx (both) / TX MCS. 8 MHz = ch44/924 MHz, US. Signal −37…−43 dBm (bench). Sampled every 10 s.

### TCP saturated — mean **15.06 Mbps**, σ 0.80, CoV **5.3%** (min 13.70 / max 17.30)
![8MHz TCP soak](https://raw.githubusercontent.com/winson3QQ/Batman/perf-8mhz-soak-charts/docs/perf/8mhz-soak/soak_8mhz_tcp.png)

### UDP 20 M offered — received **13.23 Mbps**, σ 0.23, CoV **1.8%**, loss **0.019%**
(offered 20 M but only ~13.2 arrived with ~0 loss ⇒ **sender-limited** — the small Pi can't source 20 M UDP — not link saturation; the link ceiling is the TCP figure.)
![8MHz UDP soak](https://raw.githubusercontent.com/winson3QQ/Batman/perf-8mhz-soak-charts/docs/perf/8mhz-soak/soak_8mhz_udp.png)

### Readings
- **Throughput stable across 5 min** with OTS running concurrently — no drift, no degradation.
- **CPU < 11 %** on both nodes the whole time (huge headroom; iperf + OTS together barely register).
- **RAM flat, no leak**: manet01 ≈ **1080 MB** (OTS 6 containers) vs manet02 ≈ **220 MB** (no OTS). Feeds the resource-budget question in #171.
- **RSSI −37…−43 dBm, TX MCS steady at 7** (occasional single-sample dip to 6) — solid link.
- **Bottleneck is the RF link, not CPU/OTS.**

Method: `iperf` over the mesh (br-ahwlan/bat0, not ethernet); on-node samplers for CPU (/proc/stat) / RAM (/proc/meminfo) / RSSI+MCS (`iw station dump`). Charts + tooling committed under `docs/perf/8mhz-soak/` (branch `perf-8mhz-soak-charts`). Caveat: bench range (−40 dBm best-case); 8 MHz range/robustness half still untested.

**Comment by winson3QQ** [comment]
## Throughput optimization — is there headroom? (research, on-node manet02, 8 MHz, −38 dBm bench)

Short answer: **limited, but a few concrete levers.** Aggregation is already maxed; most of the 15 Mbps-vs-31.7 gap is irreducible protocol overhead. Evidence:

### Already optimal
- **A-MPDU aggregation is ON and at the ceiling**: `Maximum RX AMPDU length 65535 bytes` (64 KB, max exponent) — `maxampdulen` is **not** the limiter. Block-ack working (TX/RX BlockAck counters healthy).
- **`expected throughput: 31.74 Mbps`** (driver's MAC estimate at MCS7/8 MHz). Real TCP goodput 15.06 Mbps ≈ **47 %** of that — typical for a half-duplex HaLow mesh with batman-adv + TCP on top. That ~53 % overhead is largely irreducible (half-duplex, per-frame airtime, batman-adv encap, TCP acks).

### Real, tunable levers (need measured experiments — one knob at a time)
1. **A-MPDU vs TBTT (#37).** Under load: `AGG crosses TBTT: 3710` (beacon_int=1000 ms → a TBTT every 1 s truncates aggregates) and `AGG TX param mismatch: 36709` (rate flapping MCS6↔7 aborts aggregates). Experiment: raise `beacon_int` (fewer TBTTs) and/or damp rate-control thrash; measure aggregate-size histogram + goodput before/after.
2. **Rate-control thrash.** TX MCS histogram under load = **MCS6 66 % / MCS7 32 %** (never 8/9). The MCS6↔7 flapping is what drives the 36 709 param-mismatches. Pinning to MCS7 (if the link sustains it) could cut aggregation breakage.
3. **256-QAM (MCS8/9) — uncertain.** Hardware advertises VHT MCS0-9 (256-QAM capable), but rate control won't select it and the link shows **7.2 % tx retries + RX FCS fail 1159 + bad-STF 146** — i.e. not clean enough. Forcing MCS8/9 will likely raise retries and *lower* goodput. Testable, but low-probability win at this link quality.
4. **Channel cleanliness.** The retry/FCS rate at −38 dBm hints at co-channel noise around 922-924 MHz. Trying another 8 MHz channel (ch12/908 or ch28/916) and comparing retry rate is a cheap experiment.

### What is NOT the bottleneck
- Not CPU (node < 11 %, chip MAC CPU 6 %), not RAM, not aggregation length, not RSSI.
- UDP goodput (13.2 Mbps) was **sender-CPU-limited** (Pi couldn't source 20 M UDP), not the link.

### Recommendation
Expect **modest** gains (single-digit %), mostly from #37 (TBTT/aggregation) and reducing rate-control flap. Suggest a controlled experiment: vary `beacon_int`, measure `AGG crosses TBTT` + goodput; then a forced-MCS7 vs auto comparison. 256-QAM is a long shot here. The big structural win remains 8 vs 4 MHz (already +65 %, done). Related: #37 (A-MPDU/TBTT), #40 (beacon_int).

**Comment by winson3QQ** [comment]
## Tuning experiments — which adjustable parameters actually move throughput? (empirical, 8 MHz, −38 dBm bench)

Ran controlled before/after sweeps (each config: set both nodes → wifi reload → verify mesh → TCP 20 s + stat deltas). **Result: no soft-tuning knob moves throughput. The only lever that worked is channel bandwidth.**

| Parameter | Adjustable? | Throughput effect |
|---|---|---|
| **BW 4→8 MHz** | yes | **+65 % (~9.5→15 Mbps) — the only real lever, structural** |
| `beacon_int` 500 / 1000 / 2000 / 4000 | yes | **none** — 14.8–15.2 Mbps (noise). beacon_int=2000 cut `AGG crosses TBTT` 263→**21** (12×) with **zero** throughput change |
| `maxampdulen` | already at max | RX A-MPDU already 65535 B (64 KB) — not the limiter |
| channel (ch28/912, ch12/908) | flaky | mesh would **not reform** on ch28/ch12 in 60 s; and retries don't track throughput anyway (see below) |
| MCS 8/9 (256-QAM) | rate-control refuses | HW advertises VHT MCS0-9 but RC stays ≤7; link shows 7 % retries + FCS/STF fails → 256-QAM won't hold; forcing it would lower goodput |
| UDP socket buffer `-w` | yes | only induces loss at the ~13.5 Mbps UDP ceiling; doesn't raise it |

### Key negative results
- **#37 (A-MPDU vs TBTT) is empirically disproven here**: cutting TBTT-aggregate-crossings 12× (via beacon_int) gave **no** throughput gain — TBTT truncation is not the bottleneck at this link.
- **Retries don't limit throughput**: across the beacon sweep tx-retries ranged 1917–2514 while throughput stayed ~15 — so a "cleaner channel" wouldn't help tput either.

### Interpretation
~15 Mbps TCP is a **robust link ceiling** for this 1-stream MM6108 at 8 MHz/MCS7. Throughput = BW × MCS × HaLow-MAC-efficiency, all near their practical ceiling; the 15-vs-31.7-"expected" gap is the intrinsic HaLow half-duplex + batman-adv + TCP tax and is not tunable away. Bonus finding: **UDP one-way (~13.5 Mbps ceiling) is *below* TCP (~15)** — the half-duplex ACK cost is small, and TCP's back-pressured byte-stream aggregates into A-MPDUs better than paced UDP (which just drops when overdriven). On aggregation-dominated HaLow, well-behaved TCP beats naive UDP.

**Takeaway: further throughput needs more bandwidth or more spatial streams (MM6108 is 1SS) — not config tuning.** All nodes reverted to ch44/8 MHz, beacon_int 1000, mesh healthy.

**Comment by winson3QQ** [comment]
**Evaluated + decision: 4 MHz is the golden default.** Full tuning matrix (this issue): the only lever on throughput is bandwidth — 8 MHz gives +65% (~9.5→15 Mbps) but at the cost of ~3 dB sensitivity / shorter range; no soft knob (beacon_int/maxampdulen/MCS/UDP-buf) moves throughput. **Decision: keep 4 MHz (ch40) as the golden default for range/robustness; 8 MHz stays an opt-in for short-range/high-SNR links.** Fleet reverted to ch40 this session (both nodes' p5 persisted to channel=40). ~15 Mbps @8 MHz was a structural HaLow half-duplex + 1SS ceiling; more throughput needs more BW or more spatial streams, not tuning. Closing.
